ir/node.h

```diff
@@ -60,12 +68,19 @@ class INode : public Util::IHasSourceInfo, public IHasDbPrint, public ICastable
     virtual cstring node_type_name() const = 0;
     virtual void validate() const {}
     virtual const Annotation *getAnnotation(cstring) const { return nullptr; }
-    /// A checked version of INode::to. A BUG occurs if the cast fails.
-    ///
-    /// A similar effect can be achieved with `&as<T>()`, but this method
-    /// produces a message that is easier to debug.
+
+    // default checkedTo implementation for nodes: just fallback to generic ICastable method
+    template <typename T>
+    const T *checkedTo(
+        typename std::enable_if_t<!has_static_type_name<T>::value> * = nullptr) const {
```

Comment:
Hmm, could you place the enable_if on the return type? I don't like the default argument much as it can confuse someone (how much easier this would be with concepts!).